echopop/live/live_data_loading.py

```diff
@@ -0,0 +1,928 @@
+    """Construct the root directory path."""
+
+    # Get the general root_directory, if present
+    if "data_root_dir" in file_configuration:
+        root_directory = file_configuration["data_root_dir"]
+    else:
+        root_directory = ""
+
+    # Get the local directory (or this may be the root directory depending on the config)
+    data_directory = file_settings["directory"]
+
+    # Return the directory path
+    if root_directory != "":
+        return "/".join([root_directory, data_directory])
+    else:
+        return data_directory
+
+
+def is_s3_path(path):
+    """Check if a path is an S3 path."""
+    return path.startswith("s3://")
+
+
+# TODO: Documentation
+def validate_data_directory(
+    file_configuration: dict, dataset: str, input_filenames: Optional[list] = None
+) -> List[Path]:
+
+    # Get the dataset file settings
+    file_settings = file_configuration["input_directories"][dataset]
+
+    # Get the data file settings and directorypath
+    directory_path = construct_directorypath(file_configuration, file_settings)
+
+    # Validate `input_filenames` input
+    if input_filenames is not None and not isinstance(input_filenames, list):
+        raise TypeError("Data loading argument `input_filenames` must be a list.")
+
+    # Format data filenames
+    if input_filenames is not None:
+        data_files = ["/".join([directory_path, filename]) for filename in input_filenames]
+
+    # Validate directories and format filepath names
+    # ---- S3 bucket
+    if is_s3_path(directory_path):
+        # ---- Validate
+        validate_s3_path(directory_path, file_configuration["storage_options"])
+        # ---- Format data files
+        if input_filenames is None:
+            data_files = []
+    # ---- Local
+    else:
+        # ---- Validate
+        validate_local_path(directory_path, file_settings)
+        # ---- Format data files
+        if input_filenames is None:
+            data_files = list(Path(directory_path).glob(f"*{'.'+file_settings['extension']}"))
+
+    # Clean the filenames
+    data_files = [
+        (
+            re.sub(r"//", r"\\", str(filename)).replace("/", "\\")
+            if not str(filename).startswith("s3://")
+            else str(filename)
+        )
+        for filename in data_files
+    ]
+
+    # Database root directory
+    database_root_directory = file_configuration["database_directory"]
+
+    # Initialize the database file
+    initialize_database(database_root_directory, file_settings)
+
+    # Drop incomplete datasets
+    if dataset == "biology":
+        data_files = validate_complete_biology_dataset(
+            data_files, directory_path, file_configuration
+        )
+
+    # Query the SQL database to process only new files (or create the db file in the first place)
+    valid_files, file_configuration["database"][dataset] = query_processed_files(
+        database_root_directory, file_settings, data_files
+    )
```

Comment:
This section deviates from the goal to validate paths as suggested by the function name.
Probably better to separate these out.